### PR TITLE
Call gtFoldExpr in morph for tier0

### PR DESCRIPTION
**Pull request by EgorBo**
`gtFoldExpr` doesn't need to be wrapped with `opts.OptimizationEnabled()` since it already checks it inside and allows some lightweight optimizations for tier0 (but not for minopts and/or debug).

Seeing some nice tier0 diffs from it.

### CODE REVIEW

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -8330,11 +8327,8 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac, bool* optA
         qmarkOp2 = oldTree->AsOp()->gtOp2->AsOp()->gtOp2;
     }
 
-    if (opts.OptimizationEnabled())
-    {
-        // Try to fold it, maybe we get lucky,
-        tree = gtFoldExpr(tree);
-    }
+    // Try to fold it, maybe we get lucky,
+    tree = gtFoldExpr(tree);
```
Comment (tannergooding) [inline src/coreclr/jit/morph.cpp:8331]: We have a similar check for hwintrinsic, right? Should that be included in this or as well?

**Review by tannergooding — APPROVED** [review]

File: src/coreclr/jit/morph.cpp (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/jit/morph.cpp:8331]: Yep, looks like those are fine as is

**Comment by jakobbotsch** [comment]
Is the TP impact worth it?

**Comment by EgorBo** [comment]
> Is the TP impact worth it?

are 0.1%-0.2% for real run collections a big deal? We currently have a problem when hot Tier0 code can stuck for 10ish seconds in tier0 due to promotion-queue problem in large apps leading to poor RPS till it reaches the steady state. Initially I also did a small clean up in OptimizationEnabled/OptimizationDisabled that improved TP by 0.15% just didn't push as part of this pr

**Comment by jakobbotsch** [comment]
> > Is the TP impact worth it?
> 
> are 0.1%-0.2% for real run collections a big deal? We currently have a problem when hot Tier0 code can stuck for 10ish seconds in tier0 due to promotion-queue problem in large apps leading to poor RPS till it reaches the steady state. Initially I also did a small clean up in OptimizationEnabled/OptimizationDisabled that improved TP by 0.15% just didn't push as part of this pr

For me it's more about that we (the JIT team) should align on philosophy of the trade-offs between CQ and TP in tier-0. We can certainly live with such small TP impact (and I wouldn't bother reverting this or anything), but in my opinion tier-0 changes need additional motivation than just CQ if they are trading off for TP. E.g. this change trades off `+0.14%` TP for `-0.06%` code size in benchmarks.run_pgo for the tier-0 contexts -- I personally wouldn't have gone for that.

**Comment by EgorBo** [comment]
To be fair, we can't even detect any visible improvements from changes like -10.0% in MinOpts in any of our benchmarks (TE startup time/time to first request). I think it's more important for Tier0 to call certain JIT-EE APIs less (because some of them may have locks, may trigger type loading etc), but we currently have no infra to measure that.

**Comment by jakobbotsch** [comment]
> To be fair, we can't even detect any visible improvements from changes like -10.0% in MinOpts in any of our benchmarks (TE startup time/time to first request).

Well, to be fair we have never done any proper study of this. The variance on this number is just so high that you need hundreds or thousands of runs with and without the change to be able to detect a JIT change in the output (couple that with the fact that large chunk of start up time is also not spent in JIT).
I definitely agree that 10% in JIT does not translate to 10% in start up time, and maybe it's easier to make up JIT regressions in other places.

> I think it's more important for Tier0 to call certain JIT-EE APIs less (because some of them may have locks, may trigger type loading etc), but we currently have no infra to measure that.

I agree -- that would be one of the "additional motivations" we can have for changes to tier-0 codegen.

**Comment by EgorBo** [comment]
Filed https://github.com/dotnet/runtime/pull/105190 - it's zero asm diffs with up to `-0.5%` TP improvements for MinOpts

**Pull request merged**